### PR TITLE
Fix SSE-C checksums

### PR DESCRIPTION
**Pull request by klauspost**
## Description

SSE-C checksums could not be decrypted, since it missed the key.

Should also s3zip files with SSE-C.

## How to test this PR?
```
λ aws s3api --debug --endpoint-url https://127.0.0.1:9001 put-object --bucket testbucket --key hello.txt --body hello.txt --sse-customer-algorithm AES256 --sse-customer-key "2teLYwr0HB84FYUxvhiTG//u3C4/RnI2qhRt6jn64K0=" --sse-customer-key-md5 "h06pENAWWRBnqFm+lHPtUA==" --no-verify-ssl --checksum-algorithm SHA256 --checksum-sha256 CKw2JCTHBUK0z6+wVutMXRw4SDyMxn2eIEHQDu3o8d4=
{
    "ETag": "\"504bad1096d7cc8817127db0e25afc56\"",
    "ChecksumSHA256": "CKw2JCTHBUK0z6+wVutMXRw4SDyMxn2eIEHQDu3o8d4=",
    "VersionId": "f6805362-a947-4f99-9096-e67cb47db11e",
    "SSECustomerAlgorithm": "AES256",
    "SSECustomerKeyMD5": "h06pENAWWRBnqFm+lHPtUA=="
}

λ aws s3api head-object --bucket testbucket --key hello.txt --endpoint-url https://localhost:9001 --checksum-mode ENABLED  --no-verify-ssl --sse-customer-algorithm AES256 --sse-customer-key "2teLYwr0HB84FYUxvhiTG//u3C4/RnI2qhRt6jn64K0=" --sse-customer-key-md5 "h06pENAWWRBnqFm+lHPtUA=="
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-07T09:08:51+00:00",
    "ContentLength": 8,
    "ChecksumSHA256": "CKw2JCTHBUK0z6+wVutMXRw4SDyMxn2eIEHQDu3o8d4=",
    "ETag": "\"504bad1096d7cc8817127db0e25afc56\"",
    "VersionId": "f6805362-a947-4f99-9096-e67cb47db11e",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "SSECustomerAlgorithm": "AES256",
    "SSECustomerKeyMD5": "h06pENAWWRBnqFm+lHPtUA==",
}

λ aws s3api get-object --bucket testbucket --key hello.txt --endpoint-url https://localhost:9001 --checksum-mode ENABLED  --no-verify-ssl --sse-customer-algorithm AES256 --sse-customer-key "2teLYwr0HB84FYUxvhiTG//u3C4/RnI2qhRt6jn64K0=" --sse-customer-key-md5 "h06pENAWWRBnqFm+lHPtUA==" out.txt
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-07T09:08:51+00:00",
    "ContentLength": 8,
    "ETag": "\"504bad1096d7cc8817127db0e25afc56\"",
    "ChecksumSHA256": "CKw2JCTHBUK0z6+wVutMXRw4SDyMxn2eIEHQDu3o8d4=",
    "VersionId": "f6805362-a947-4f99-9096-e67cb47db11e",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "SSECustomerAlgorithm": "AES256",
    "SSECustomerKeyMD5": "h06pENAWWRBnqFm+lHPtUA==",
}
```

Works with already uploaded objects.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by shtripat** [comment]
if site replication enabled, object gets replicated fine but checksum is not getting replicated as below
```
$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --checksum-mode ENABLED  --no-verify-ssl --sse-customer-algorithm AES256 --sse-customer-key "2teLYwr0HB84FYUxvhiTG//u3C4/RnI2qhRt6jn64K0=" --sse-customer-key-md5 "h06pENAWWRBnqFm+lHPtUA==" 
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-07T12:03:35+00:00",
    "ContentLength": 16,
    "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY=",
    "ETag": "\"9950c98022b087757a0439f55e3bfd7f\"",
    "VersionId": "f779e5fe-f6d5-4e4f-9130-a880d9e2fb79",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "SSECustomerAlgorithm": "AES256",
    "SSECustomerKeyMD5": "h06pENAWWRBnqFm+lHPtUA==",
    "ReplicationStatus": "COMPLETED"
}

$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9010 --checksum-mode ENABLED  --no-verify-ssl --sse-customer-algorithm AES256 --sse-customer-key "2teLYwr0HB84FYUxvhiTG//u3C4/RnI2qhRt6jn64K0=" --sse-customer-key-md5 "h06pENAWWRBnqFm+lHPtUA==" 
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-07T12:03:35+00:00",
    "ContentLength": 16,
    "ETag": "\"9950c98022b087757a0439f55e3bfd7f\"",
    "VersionId": "f779e5fe-f6d5-4e4f-9130-a880d9e2fb79",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "SSECustomerAlgorithm": "AES256",
    "SSECustomerKeyMD5": "h06pENAWWRBnqFm+lHPtUA==",
    "ReplicationStatus": "REPLICA"
}
```

looks a small piece missing or need to be added the allowed headers to replicate checksum?

**Comment by klauspost** [comment]
@shtripat Yes - it will require transferring the encrypted checksum. This does not do that, it only fixes it locally.

File: cmd/encryption-v1.go [hunk 1]
```diff
@@ -1077,13 +1077,17 @@ func metadataEncrypter(key crypto.ObjectKey) objectMetaEncryptFn {
...
 		if len(input) == 0 {
 			return input, nil
 		}
-
-		key, err := decryptObjectMeta(nil, o.Bucket, o.Name, o.UserDefined)
+		var key []byte
+		if k, err := crypto.SSEC.ParseHTTP(h); err == nil {
+			key = k[:]
+		}
+		fmt.Println("GOT SSE-C KEY", key)
```
Comment (harshavardhana) [inline cmd/encryption-v1.go]:
```suggestion
```

**Review by harshavardhana — APPROVED** [review]

**Review by poornas — APPROVED** [review]

**Comment by klauspost** [comment]
@harshavardhana https://github.com/minio/minio-go/pull/1970 is needed. I will open a separate PR.

**Pull request merged**